### PR TITLE
G2 wrap-up

### LINKED ISSUE #8357: G2 wrap-up
- [ ] Add a ChangeLog entry for G2.
- [ ] Update user documentation: `docs/driver-only-builds.md`.
- [ ] Update any strategy document (`docs/architecture/psa-migration/*.md`) that needs updating.
- (List to be completed if and when we notice any other small enough item - larger items should be their own issue.)
- [ ] Update our copy of the TF-M config to no longer defined `CIPHER_C`?

**Depends on:** every other G2 task in this EPIC :)

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #8357

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required
- [x] **tests** provided

### CODE REVIEW

File: ChangeLog.d/8357.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Features
+   * It is now possible to have AEADs support (CCM, GCM and ChaChaPoly) without
+     MBEDTLS_CIPHER_C. This holds both for the builtin suport (MBEDTLS_CCM_C,
+     MBEDTLS_GCM_C and MBEDTLS_CHACHAPOLY_c) as well as the PSA one
```
Comment (mpg) [inline ChangeLog.d/8357.txt:4]: Typo: `_c` -> `_C`.

File: ChangeLog.d/8357.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+     (PSA_WANT_ALG_CCM, PSA_WANT_ALG_GCM, PSA_WANT_ALG_CHACHA20_POLY1305).
+     On the PSA side this means that it is possible to enable
+     MBEDTLS_PSA_CRYPTO_C without MBEDTLS_CIPHER_C if any of the
```
Comment (mpg) [inline ChangeLog.d/8357.txt:7]: I think you mean if none of the unauthenticated  ciphers is enabled?

File: docs/driver-only-builds.md [hunk 3]
```diff
@@ -240,3 +241,26 @@ removing builtin support (i.e. `MBEDTLS_DHM_C`).
 ### Limitations
 Support for deterministic derivation of a DH keypair
 (i.e. `PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_DERIVE`) is not supported.
+
+AEADs
+-----
+
+It is possible to have all AEADs operations provided only by a driver.
```
Comment (mpg) [inline docs/driver-only-builds.md:248]: I don't disagree, but that's G3 and this is supposed to be G2 wrap-up :)

Maybe skip this commit for now and work on the text in the G3-G4 wrap-up issue instead? The situation is actually not that easy to describe in a way that's both complete and readable.

File: docs/driver-only-builds.md [hunk 4]
```diff
@@ -240,3 +241,26 @@ removing builtin support (i.e. `MBEDTLS_DHM_C`).
...
+
+More precisely you can:
+- enable desired PSA algorithm(s) and key type(s):
+  - `PSA_WANT_ALG_[CCM|GCM]` with `PSA_WANT_KEY_TYPE_[AES|ARIA|CAMELLIA]`
+  - `PSA_WANT_ALG_CHACHA20_POLY1305` with `PSA_WANT_KEY_TYPE_CHACHA20`;
+- enable `MBEDTLS_PSA_ACCEL_xxx` symbol(s) which correspond to the
+  `PSA_WANT_xxx` of the previous step;
+- disable builtin support of `MBEDTLS_[CCM|GCM|CHACHAPOLY]_C` algorithms and
+  key types `MBEDTLS_[AES|ARIA|CAMELLIA|CHACHA20]_C` for AEADs which are
```
Comment (mpg) [inline docs/driver-only-builds.md:257]: Note: for ChachaPoly you can also disable `MBEDTLS_POLY1305_C`.

File: docs/driver-only-builds.md [hunk 5]
```diff
@@ -240,3 +241,26 @@ removing builtin support (i.e. `MBEDTLS_DHM_C`).
...
+  accelerated.
+
+In such a build all AEADs operations requested through the PSA Crypto API
+(including those in TLS and X.509) will be performed by the provided driver.
+Of course direct calls to the disabled builtin modules
+(ex: `mbedtls_ccm_init()`, etc) won't be possible.
+
+If no other non-authenticated cipher is required, it is also possible to
```
Comment (mpg) [inline docs/driver-only-builds.md:265]: Wording: "If no unauthenticated cipher" instead. (Unauthenticated ciphers are not AEADs, while "other" suggests they are.)

File: docs/driver-only-builds.md [hunk 6]
```diff
@@ -240,3 +241,26 @@ removing builtin support (i.e. `MBEDTLS_DHM_C`).
...
+- enable `MBEDTLS_PSA_ACCEL_xxx` symbol(s) which correspond to the
+  `PSA_WANT_xxx` of the previous step;
+- disable builtin support of `MBEDTLS_[CCM|GCM|CHACHAPOLY]_C` algorithms and
+  key types `MBEDTLS_[AES|ARIA|CAMELLIA|CHACHA20]_C` for AEADs which are
+  accelerated.
+
+In such a build all AEADs operations requested through the PSA Crypto API
+(including those in TLS and X.509) will be performed by the provided driver.
+Of course direct calls to the disabled builtin modules
+(ex: `mbedtls_ccm_init()`, etc) won't be possible.
```
Comment (mpg) [inline docs/driver-only-builds.md:263]: Also, once you've disabled `AES_C` you won't get PEM descrytion, or PKCS5/15 or their use in in PKparse. (And perhaps others I'm forgetting, we should double-check what's disabled in the `all.sh` component and what's ignored in `analyze_coverage.py`.)

So, I think we need to change the structure like:
1. if all AEAD algs are accelerated, you can disable the AEAD modules (TODO: list them) and everything will work.
2. if additionally you don't need unauthenticated ciphers, you can disable `AES_C` etc and also `CIPHER_C` and everything will work except for PEM etc. (TODO: complete list - for now this even includes `CTR_DRBG_C`).

File: configs/config-tfm.h [hunk 7]
```diff
@@ -45,6 +45,11 @@
 #undef MBEDTLS_PLATFORM_STD_EXIT_SUCCESS
 #undef MBEDTLS_PLATFORM_STD_EXIT_FAILURE
 
+/* CCM is the only cipher/AEAD enabled in TF-M configuration files, but it
+ * does not need CIPHER_C to be enabled, so we can disabled it in order
```
Comment (mpg) [inline configs/config-tfm.h:49]: Typo: disableD

**Review by mpg — CHANGES_REQUESTED** [review]
Apart from a few typos/wording improvements, most of my comments about the part that's about G3, so I suggest we leave it out for now to get this merged, and we can pick it up later when we do G3 wrap-up.

**Comment by valeriosetti** [comment]
@mpg as you correctly pointed out in [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8599#discussion_r1415388595) it's very likely that `driver-only-builds.md` was not the right target of this PR, because this one is the wrap up of G2 which does not involve driver's acceleration. However, since I already started that section, I thought that instead of just dropping the commit, it was worth to add a disclaimer stating that it will be updated soon by another PR and then trying reshape it following your inputs of the following comments.
This seemed reasonable to me, but in case you don't like this approach just let me know and I will remove that change/commit(s).

**Review by mpg — APPROVED** [review]
LGTM.

(I'm not fully happy with the text in driver-only-builds.md yet, but I'm leaving that for later.)

**Review by d3zd3z — APPROVED** [review]

**Pull request merged**